docs/src/literate/ns_vs_diffeq.jl

```diff
@@ -0,0 +1,463 @@
+
+    # Before starting the nonlinear solve we have to set the time correctly.
+    # Note that ch is a global variable for now.
+    #+
+    update!(ch, cache.tstep);
+
+    # The update of u takes uprev + z or tmp + z most of the time, so we have
+    # to enforce Dirichlet BCs here. Note that these mutations may break the
+    # error estimators.
+    #+
+    apply!(uprev, ch)
+    apply!(tmp, ch)
+    apply_zero!(z, ch);
+
+    nothing
+end;
+
+# For the linear equations we can cleanly integrate with the linear solver
+# interface provided by the DifferentialEquations ecosystem. We use a direct
+# solver for simplicity, altough it comes with some issues. Implementing
+# GMRES with efficient preconditioner is left open for future work.
+mutable struct FerriteLinSolve{CH,F}
+    ch::CH
+    factorization::F
+    A
+end
+FerriteLinSolve(ch) = FerriteLinSolve(ch,lu,nothing)
+function (p::FerriteLinSolve)(::Type{Val{:init}},f,u0_prototype)
+    FerriteLinSolve(ch)
+end
+function (p::FerriteLinSolve)(x,A,b,update_matrix=false;reltol=nothing, kwargs...)
+    if update_matrix
+        ## Apply Dirichlet BCs
+        apply_zero!(A, b, p.ch)
+        ## Update factorization
+        p.A = p.factorization(A)
+    end
+    ldiv!(x, p.A, b)
+    apply_zero!(x, p.ch)
+    return nothing
+end;
+
+# DifferentialEquations assumes dense matrices by default, which is not
+# feasible for semi-discretization of finize element models. We communicate
+# that a sparse matrix with specified pattern should be utilized through the
+# `jac_prototyp` argument. Additionally, we have to provide the mass matrix.
+# To apply the nonlinear portion of the Navier-Stokes problem we simply hand
+# over the dof handler to the right hand side as a parameter in addition to
+# the pre-assembled linear part (which is time independent) to save some
+# runtime.
+jac_sparsity = sparse(K)
+function navierstokes!(du,u,p,t)
+    K,dh,cellvalues_v = p
+    du .= K * u
+
+    n_basefuncs = getnbasefunctions(cellvalues_v)
+
+    ## Nonlinear contribution
+    for cell in CellIterator(dh)
+        Ferrite.reinit!(cellvalues_v, cell)
+        ## Trilinear form evaluation
+        all_celldofs = celldofs(cell)
+        v_celldofs = all_celldofs[dof_range(dh, :v)]
+        v_cell = u[v_celldofs]
+        for q_point in 1:getnquadpoints(cellvalues_v)
+            dΩ = getdetJdV(cellvalues_v, q_point)
+            ∇v = function_gradient(cellvalues_v, q_point, v_cell)
+            v = function_value(cellvalues_v, q_point, v_cell)
+            for j in 1:n_basefuncs
+                φⱼ = shape_value(cellvalues_v, q_point, j)
+                # Note that the order the gradient term is now on the left, which is the correct thing to do here.
+                # It can be quickly shown through index notation.
+                # ```math
+                # [(v \cdot \nabla) v]_i = v_j \partial_j v_i = \partial_j v_i v_j = (\nabla v) v
```

Comment:
```suggestion
                # [(v \cdot \nabla) v] = v_i \partial_i v_j = \partial_i v_i v_j = (\nabla v) v
```

Otherwise you'd have to write 
```md
               # [(v \cdot \nabla) v]_{ij} = v_i \partial_i v_j = \partial_i v_i v_j = (\nabla v) v
```

Because in tensor notation you only apply indices to a bracket if you deconstruct every inner running index, in order to show the order of the tensor. At least this is what is done in the mechanics community and "canonical" in the sense of tensor notation. Sometimes, also utilized to denote a derivative with running index to a bracket by e.g. `(v_i v_j)_{,k}` would be the derivative with running index `k` to the inner arguments. Nonetheless, the indices should start first with i and secondly with j, by that you get naturally the correct nabla formulation, you can also be very explicit by 

```md
               # [(v \cdot \nabla) v] = v_i \boldsymbol{e}_i \cdot \partial_j \boldsymbol{e}_j v_k \boldsymbol{e}_k = v_i \partial_j \delta_{ij} v_k \boldsymbol{e}_k = v_i \partial_i \boldsymbol{e}_k v_k \boldsymbol{e}_k = \partial_i v_i v_j = (\nabla v) v
```